src/server/scripts/Outland/HellfireCitadel/BloodFurnace/instance_blood_furnace.cpp

```diff
@@ -169,28 +170,37 @@ class instance_blood_furnace : public InstanceMapScript
 
             void ResetPrisons()
             {
-                PrisonerCounter5 = PrisonersCell5.size();
                 ResetPrisoners(PrisonersCell5);
+                PrisonerCounter5 = PrisonersCell5.size();
                 HandleGameObject(PrisonCellGUIDs[DATA_PRISON_CELL5 - DATA_PRISON_CELL1], false);
 
-                PrisonerCounter6 = PrisonersCell6.size();
                 ResetPrisoners(PrisonersCell6);
+                PrisonerCounter6 = PrisonersCell6.size();
                 HandleGameObject(PrisonCellGUIDs[DATA_PRISON_CELL6 - DATA_PRISON_CELL1], false);
 
-                PrisonerCounter7 = PrisonersCell7.size();
                 ResetPrisoners(PrisonersCell7);
+                PrisonerCounter7 = PrisonersCell7.size();
                 HandleGameObject(PrisonCellGUIDs[DATA_PRISON_CELL7 - DATA_PRISON_CELL1], false);
 
-                PrisonerCounter8 = PrisonersCell8.size();
                 ResetPrisoners(PrisonersCell8);
+                PrisonerCounter8 = PrisonersCell8.size();
                 HandleGameObject(PrisonCellGUIDs[DATA_PRISON_CELL8 - DATA_PRISON_CELL1], false);
             }
 
-            void ResetPrisoners(GuidSet const& prisoners)
+            void ResetPrisoners(GuidSet& prisoners)
             {
+                std::list<ObjectGuid> deadPrisoners;
                 for (GuidSet::const_iterator i = prisoners.begin(); i != prisoners.end(); ++i)
-                    if (Creature* prisoner = instance->GetCreature(*i))
+                    if (Creature * prisoner = instance->GetCreature(*i))
```

Comment:
this loop could be changed to store "i" in the loop as another variable (currentItr ?), increment "i" while the iterator is still valid, then process currentItr and call prisoners.erase(currentItr) if the prisoner is not alive.
This way you don't need another container (unless there is another reason not to modify it)

Reply:
I tried doing something of sort, but I end up with access violation exception (or sort of) leaving log like "unable to advance iterator" or something like that. I try to implement what you mean, still new to c++ language constructs but it is a learning curve so I will get there sooner or later.

Reply:
I usually check http://www.cplusplus.com/reference/ like http://www.cplusplus.com/reference/set/set/erase/ or https://en.cppreference.com/w/ like https://en.cppreference.com/w/cpp/container/set/erase to check how to avoid invalidating an iterator.

There might be some nice tutorials online about iterators, a short tl;dr is if you erase the iterator from a container, the iterator will most likely be invalid and you can't increment it anymore. If you store the current iterator into another variable, then you have 2 different iterators, so you can increment (safely) one and use the other one as parameter for erase().

I would suggest to trying to solve it by yourself, that's the best way to learn, but let me know anyway if you prefer me to post a snippet of the code already updated.